tests/enrol_advanced_testcase_test.php

```diff
@@ -0,0 +1,256 @@
+          if (empty($instanceid))
+          {
+            $instanceid = $evento_plugin->add_instance($course);
+          }
+        }
+        if (!empty($instanceid))
+        {
+          // Do additional config of instance if needed.
+          ($instanceid);
+        }
+      }
+
+      $record = new stdClass();
+      $record->id = 246003;
+      $record->customint1 = 0;
+      $DB->update_record('enrol', $record, $bulk=false);
+      /*Enrol Users into courses*/
+      $enrol->user_sync($trace, $courseid =null);
+      /*Get user enrolment record to count enrolments*/
+      $this->get_enroled_user($this->course1->id);
+      $this->assertEquals($this->enrolments, 33, "Einschreibungen");
+    }
+    public function test_update_student_enrolment()
+    {
+      global $DB;
+      $this->resetAfterTest(false);
+      $eventoenrolstate = 20215;
+      $eventopersonid =  141703;
+      /*Get the user records*/
+      $person = $this->locallib->get_users_by_eventoid_exposed($eventopersonid, $isstudent=null);
+      $user = reset($person);
+      /*Get course settings*/
+      $course = $DB->get_record('course', array('idnumber'=>'mod.mmpAUKATE1.HS18_BS.001'));
+      $this->get_enroled_user($course->id);
+
+      /*Delete a user to re-enrol after*/
+      $DB->delete_records('user_enrolments', array('userid'=> current($person)->id));
+      $DB->get_record('user_enrolments', array('userid'=> current($person)->id));
+      $this->assertEquals($DB->count_records('user_enrolments', array('enrolid'=> $this->user_enrolment->id)), 32);
+      /*Get enrolment instance of course*/
+      $instance = $DB->get_record('enrol', array('id' => $this->user_enrolment->id));
+      /*re-enrol user*/
+      $this->locallib->update_student_enrolment_exposed($eventopersonid, $eventoenrolstate, $instance);
+      /*Count users*/
+      $this->assertEquals($DB->count_records('user_enrolments', array('enrolid'=> $this->user_enrolment->id)), 33);
+    }
+    public function test_enrol_teacher()
+    {
+      sleep(20);
+      global $DB;
+      $eventopersonid =  117828;
+      $this->resetAfterTest(false);
+      /*Get the teacher record*/
+      $person_teacher = $this->locallib->get_users_by_eventoid_exposed($eventopersonid, $isstudent=null);
+      /*Get the course settings*/
+      $course = $DB->get_record('course', array('idnumber'=>'mod.mmpAUKATE1.HS18_BS.001'));
+      $this->get_enroled_user($course->id);
+      $user = $DB->get_record('user_enrolments', array('userid'=>current($person_teacher)->id));
+      /*Delete the teacher to re-enroll after*/
+      $id = $DB->delete_records('user_enrolments', array('userid'=>current($person_teacher)->id));
+      $this->assertEquals($DB->count_records('user_enrolments', array('enrolid'=> $this->user_enrolment->id)), 32);
+      /**Get the course instance*/
+      $instance = $DB->get_record('enrol', array('id'=>$this->user_enrolment->id));
+      /*Re-enroll the teacher*/
```

Comment:
Ein Test besteht typischerweise aus 3 Abschnitten:
1. Testumgebung bereitstellen.
2. Methode aufrufen, die getestet werden soll.
3. Prüfen, ob der Aufruf erfolgreich war.

Man sieht in dieser Methode gar nicht wirklich, wo was ist.
Abschnitt 2 ist:
$teacher = $this->locallib->enrol_teacher_exposed($eventopersonid, $instance);
Abschnitt 3 ist.
$this->assertEquals($DB->count_records('user_enrolments', array('enrolid'=> $this->user_enrolment->id)), 33);
Alles andere gehört zu Abschnitt 1.
Den Inhalt müsste man gut zusammenfassen.
Sonst finde ich es unübersichtlich.